library/modules/EventManager.cpp

```diff
@@ -527,12 +533,18 @@ static void manageJobCompletedEvent(color_ostream& out) {
     int32_t tick0 = eventLastTick[EventType::JOB_COMPLETED];
     int32_t tick1 = df::global::world->frame_counter;
 
-    multimap<Plugin*,EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
-    unordered_map<int32_t, df::job*> nowJobs;
-    for ( df::job_list_link* link = &df::global::world->jobs.list; link != nullptr; link = link->next ) {
-        if ( link->item == nullptr )
-            continue;
-        nowJobs.emplace(link->item->id, link->item);
+    multimap<Plugin*, EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
+    std::unordered_map<int32_t, JobCompleteData> nowJobs;
+    nowJobs.reserve(prevJobs.size());
+    for (df::job_list_link* link = &df::global::world->jobs.list; link != nullptr; link = link->next) {
+        if (link->item != nullptr) {
+            df::job& job = *link->item;
+            // Can we pack these into a single integer?
+            nowJobs[job.id] = { job.completion_timer, job.flags.bits.repeat };
+            if (savedJobs.find(job.id) == savedJobs.end()) {
+                savedJobs.emplace(job.id, Job::cloneJobStruct(&job, true));
+            }
```

Comment:
If I understand correctly, this will clone jobs only once, namely when they are encountered the first time. However, at this point items and workers will, for the most part, not yet be attached to the job.

Maybe it makes more sense to only clone jobs that have a worker attached to them. If a job disappears without ever having been picked up a worker, it most likely was cancelled. And I don't think any jobs can successfully complete the same tick they get picked up by a worker.

Reply:
I *think* that's true that jobs can't be completed in the same tick that they get a worker, even with `fastdwarf`, but we should also be able to handle the case where we happen to see the job in the same tick that it gets completed. This can happen if all the registered handlers have a non-zero refresh rate.

Reply:
Yeah, I wasn't sure about it copying only when first seen. It shouldn't hurt much to re-copy (or only some bits that changed) when a worker is added or items changed.

Tick1 is almost always greater than tick0 from my observations. I figured there must've been a good reason for it not to handle same tick completions so I left that as-is. Will try it with 0 delta.